src/main/java/org/testng/internal/TestResult.java

```diff
@@ -320,4 +336,100 @@ public boolean wasRetried() {
+        //method. So lets add everything.
+        skippedDueTo.add(skippedConfig.getMethod());
+      }
+    }
+    if (!skippedDueTo.isEmpty()) {
+      //If we found atleast one skipped due to reason, then its time to return back.
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    //Looks like we didnt have any configuration failures. So some upstream method perhaps failed.
+    List<String> upstreamMethods = Arrays.asList(m_method.getMethodsDependedUpon());
+    if (upstreamMethods.isEmpty()) {
+      //Maybe group dependencies exist ?
+      List<String> upstreamGroups = Arrays.asList(m_method.getGroupsDependedUpon());
+      if (upstreamGroups.isEmpty()) {
+        return Collections.unmodifiableList(skippedDueTo);
+      }
+      List<ITestResult> allFailures = Lists.newArrayList(m_context.getFailedTests().getAllResults());
+      allFailures.addAll(m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+      skippedDueTo = allFailures.stream()
+          .map(ITestResult::getMethod)
+          .filter(each -> {
+            List<String> currentMethodGroups = Arrays.asList(each.getGroups());
+            List<String> interection = Lists.intersection(upstreamGroups, currentMethodGroups);
+            return !interection.isEmpty();
+          }).collect(Collectors.toList());
+
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+
+    //So we have dependsOnMethod failures
+    List<ITestResult> allfailures = Lists.newArrayList(m_context.getFailedTests().getAllResults());
+    allfailures.addAll(m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+    skippedDueTo = allfailures.stream()
+        .map(ITestResult::getMethod)
+        .filter(each -> upstreamMethods.contains(each.getQualifiedName()))
+        .collect(Collectors.toList());
+    return Collections.unmodifiableList(skippedDueTo);
+  }
+
+  private static boolean isGlobalFailure(ITestResult result) {
+    ITestNGMethod m = result.getMethod();
+    return m.isBeforeTestConfiguration() || m.isBeforeSuiteConfiguration();
+  }
+
+  private boolean isRelated(ITestResult result) {
+    ITestNGMethod m = result.getMethod();
+    if (!m.isBeforeClassConfiguration() && !m.isBeforeMethodConfiguration()) {
+      return false;
+    }
+    Object current = this.getInstance();
+    Object thatObject = result.getInstance();
+    return current.getClass().isAssignableFrom(thatObject.getClass()) ||
+        thatObject.getClass().isAssignableFrom(current.getClass());
+  }
+
+  private boolean belongToSameGroup(ITestResult result) {
+    ITestNGMethod m = result.getMethod();
+    if (!m.isBeforeGroupsConfiguration()) {
+      return false;
+    }
+    String[] mygroups = this.m_method.getGroups();
+    if (mygroups.length == 0) {
+      return false;
+    }
+    List<String> cfgMethodGroups = Arrays.asList(result.getMethod().getGroups());
+    if (cfgMethodGroups.isEmpty()) {
+      return false;

```

Comment:
Maybe we can avoid the list conversion in that case.

Reply:
fixed